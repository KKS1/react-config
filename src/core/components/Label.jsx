import React, { Component } from "react";

const Label = ({ schema, label, id }) => {
  return !schema && label ? <label htmlFor={id}>{label}</label> : "";
};

export default Label;
