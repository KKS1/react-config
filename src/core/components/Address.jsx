import React, { Component } from "react";
import styled from "styled-components";
import Label from "./Label";
import LocationAutoComplete from "./LocationAutoComplete";
import TextField from "./TextField";
import Dropdown from "./Dropdown";

const STATES = [
  "",
  "AK",
  "AL",
  "AZ",
  "AR",
  "CA",
  "CO",
  "CT",
  "DE",
  "DC",
  "FL",
  "GA",
  "HI",
  "ID",
  "IL",
  "IN",
  "IA",
  "KS",
  "KY",
  "LA",
  "ME",
  "MD",
  "MA",
  "MI",
  "MN",
  "MS",
  "MO",
  "MT",
  "NE",
  "NV",
  "NH",
  "NJ",
  "NM",
  "NY",
  "NC",
  "ND",
  "OH",
  "OK",
  "OR",
  "PA",
  "RI",
  "SC",
  "SD",
  "TN",
  "TX",
  "UT",
  "VT",
  "VA",
  "WA",
  "WV",
  "WI",
  "WY"
];

const dropdownOptions = {
  enumOptions: STATES.map(state => ({
    value: state,
    label: state ? state : "Please select one"
  }))
};

const zipTextFieldOptions = {
  inputProps: { className: "w-50" }
};

const addressProps = {
  ADDRESS_1: "address1",
  ADDRESS_2: "address2",
  CITY: "city",
  STATE: "state",
  ZIP: "zip"
};

const AddressContainer = styled.div`
  & > * {
    margin-bottom: 1rem;
  }
`;

export default class Address extends Component {
  static defaultProps = {
    label: "Address",
    placeholder: "",
    name: "addressWidget",
    onChange: () => {}
  };

  constructor(props) {
    super(props);
  }

  componentDidMount() {}

  propogateOnChange = (prop, value) => {
    this.props.onChange({ ...this.props.formData, [prop]: value });
  };

  changeHandler = (prop, val) => {
    this.propogateOnChange(prop, val);
  };

  placeChangeHandler = ({ formatted_address }) => {
    if (!formatted_address) {
      return;
    }

    const placesArr = formatted_address.split(", ");
    const [country, stateAndZip, city, ...rest] = placesArr.reverse();
    const address1 = rest && rest.join(", ");
    const [state, zip] = stateAndZip.split(" ");

    const updatedDetails = {
      address1,
      city,
      state,
      zip
    };

    Object.keys(updatedDetails).forEach(prop =>
      this.changeHandler(prop, updatedDetails[prop])
    );
  };

  render() {
    const { placeholder, name, value, formData, idSchema } = this.props;
    const { address1, address2, city, state, zip } = formData;
    const { $id: id } = idSchema;

    return (
      <AddressContainer>
        <LocationAutoComplete
          id={`${id}_${addressProps.ADDRESS_1}`}
          label="Address"
          value={address1}
          onChange={val => this.changeHandler(addressProps.ADDRESS_1, val)}
          onPlaceChange={this.placeChangeHandler}
        />

        <TextField
          id={`${id}_${addressProps.ADDRESS_2}`}
          label="Address Line 2"
          value={address2}
          onChange={val => this.changeHandler(addressProps.ADDRESS_2, val)}
        />

        <TextField
          id={`${id}_${addressProps.CITY}`}
          label="City"
          value={city}
          onChange={val => this.changeHandler(addressProps.CITY, val)}
        />

        <Dropdown
          id={`${id}_${addressProps.STATE}`}
          label="State"
          customClasses="w-50"
          options={dropdownOptions}
          value={state}
          onChange={val => this.changeHandler(addressProps.STATE, val)}
        />

        <TextField
          id={`${id}_${addressProps.ZIP}`}
          label="Zip"
          options={zipTextFieldOptions}
          value={zip}
          onChange={val => this.changeHandler(addressProps.ZIP, val)}
        />
      </AddressContainer>
    );
  }
}
