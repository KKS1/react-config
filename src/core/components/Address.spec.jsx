import React, { Component } from "react";
import Label from "./Label";
import Address from "./Address";

describe("Address", () => {
  test("render_Address", () => {
    const props = {
      label: "Address Label",
      formData: {},
      idSchema: { $id: "formOne_AddressOne" }
    };

    const wrapper = shallow(<Address {...props} />);
    const instance = wrapper.instance();
    expect(instance).toMatchSnapshot();
  });

  test("changeHandler_should_call_propogateOnChange", () => {
    const props = {
      label: "Address Label",
      formData: {},
      idSchema: { $id: "formOne_AddressOne" },
      onChange: () => {}
    };

    const wrapper = shallow(<Address {...props} />);
    const instance = wrapper.instance();

    jest.spyOn(instance, "propogateOnChange");

    instance.changeHandler("city", "SF");

    expect(instance.propogateOnChange).toHaveBeenCalledWith("city", "SF");
  });

  test("propogateOnChange_should_call_propOnChange", () => {
    let addressObj;
    const props = {
      label: "Address Label",
      formData: {},
      idSchema: { $id: "formOne_AddressOne" },
      onChange: _addressObj => (addressObj = _addressObj)
    };

    const wrapper = shallow(<Address {...props} />);
    const instance = wrapper.instance();

    instance.changeHandler("city", "SF");

    expect(addressObj).toBeDefined();
    expect(addressObj).toEqual({ city: "SF" });
  });

  test("placeChangeHandler_should_call_propOnChange_with_addressObject", () => {
    let addressObj;
    const props = {
      label: "Address Label",
      formData: {},
      idSchema: { $id: "formOne_AddressOne" },
      onChange: _addressObj => (addressObj = { ...addressObj, ..._addressObj })
    };

    const wrapper = shallow(<Address {...props} />);
    const instance = wrapper.instance();

    instance.placeChangeHandler({});

    expect(addressObj).not.toBeDefined();

    instance.placeChangeHandler({
      formatted_address: "303 2nd St, San Francisco, CA 94107, USA"
    });

    expect(addressObj).toBeDefined();
    expect(addressObj).toEqual({
      address1: "303 2nd St",
      city: "San Francisco",
      state: "CA",
      zip: "94107"
    });
  });

  test("change_simulations", () => {
    let addressObj;
    const props = {
      formData: {},
      idSchema: { $id: "formOne_AddressOne" },
      onChange: _addressObj => (addressObj = { ...addressObj, ..._addressObj })
    };

    const wrapper = shallow(<Address {...props} />);
    const instance = wrapper.instance();

    wrapper.find("LocationAutoComplete").simulate("change", "303 2nd St");

    wrapper
      .find("TextField")
      .at(0)
      .simulate("change", "Suite 325");

    wrapper
      .find("TextField")
      .at(1)
      .simulate("change", "San Francisco");

    wrapper.find("Dropdown").simulate("change", "CA");

    wrapper
      .find("TextField")
      .at(2)
      .simulate("change", "94107");

    expect(addressObj).toBeDefined();

    expect(addressObj).toEqual({
      address1: "303 2nd St",
      address2: "Suite 325",
      city: "San Francisco",
      state: "CA",
      zip: "94107"
    });
  });
});
