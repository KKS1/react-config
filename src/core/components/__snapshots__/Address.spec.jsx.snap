// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Address render_Address 1`] = `
Address {
  "changeHandler": [Function],
  "context": Object {},
  "placeChangeHandler": [Function],
  "propogateOnChange": [Function],
  "props": Object {
    "formData": Object {},
    "idSchema": Object {
      "$id": "formOne_AddressOne",
    },
    "label": "Address Label",
    "name": "addressWidget",
    "onChange": [Function],
    "placeholder": "",
  },
  "refs": Object {},
  "setState": [Function],
  "state": null,
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_didScheduleRenderPhaseUpdate": false,
      "_dispatcher": Object {
        "readContext": [Function],
        "useCallback": [Function],
        "useContext": [Function],
        "useDebugValue": [Function],
        "useEffect": [Function],
        "useImperativeHandle": [Function],
        "useLayoutEffect": [Function],
        "useMemo": [Function],
        "useReducer": [Function],
        "useRef": [Function],
        "useState": [Function],
      },
      "_element": <Address
        formData={Object {}}
        idSchema={
          Object {
            "$id": "formOne_AddressOne",
          }
        }
        label="Address Label"
        name="addressWidget"
        onChange={[Function]}
        placeholder=""
      />,
      "_firstWorkInProgressHook": null,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_isReRender": false,
      "_newState": null,
      "_numberOfReRenders": 0,
      "_renderPhaseUpdates": null,
      "_rendered": <ForwardRef>
        <LocationAutoComplete
          id="formOne_AddressOne_address1"
          label="Address"
          onChange={[Function]}
          onPlaceChange={[Function]}
          placeholder=""
        />
        <TextField
          id="formOne_AddressOne_address2"
          label="Address Line 2"
          name="textField"
          onChange={[Function]}
          placeholder=""
        />
        <TextField
          id="formOne_AddressOne_city"
          label="City"
          name="textField"
          onChange={[Function]}
          placeholder=""
        />
        <Dropdown
          customClasses="w-50"
          id="formOne_AddressOne_state"
          label="State"
          onChange={[Function]}
          options={
            Object {
              "enumOptions": Array [
                Object {
                  "label": "Please select one",
                  "value": "",
                },
                Object {
                  "label": "AK",
                  "value": "AK",
                },
                Object {
                  "label": "AL",
                  "value": "AL",
                },
                Object {
                  "label": "AZ",
                  "value": "AZ",
                },
                Object {
                  "label": "AR",
                  "value": "AR",
                },
                Object {
                  "label": "CA",
                  "value": "CA",
                },
                Object {
                  "label": "CO",
                  "value": "CO",
                },
                Object {
                  "label": "CT",
                  "value": "CT",
                },
                Object {
                  "label": "DE",
                  "value": "DE",
                },
                Object {
                  "label": "DC",
                  "value": "DC",
                },
                Object {
                  "label": "FL",
                  "value": "FL",
                },
                Object {
                  "label": "GA",
                  "value": "GA",
                },
                Object {
                  "label": "HI",
                  "value": "HI",
                },
                Object {
                  "label": "ID",
                  "value": "ID",
                },
                Object {
                  "label": "IL",
                  "value": "IL",
                },
                Object {
                  "label": "IN",
                  "value": "IN",
                },
                Object {
                  "label": "IA",
                  "value": "IA",
                },
                Object {
                  "label": "KS",
                  "value": "KS",
                },
                Object {
                  "label": "KY",
                  "value": "KY",
                },
                Object {
                  "label": "LA",
                  "value": "LA",
                },
                Object {
                  "label": "ME",
                  "value": "ME",
                },
                Object {
                  "label": "MD",
                  "value": "MD",
                },
                Object {
                  "label": "MA",
                  "value": "MA",
                },
                Object {
                  "label": "MI",
                  "value": "MI",
                },
                Object {
                  "label": "MN",
                  "value": "MN",
                },
                Object {
                  "label": "MS",
                  "value": "MS",
                },
                Object {
                  "label": "MO",
                  "value": "MO",
                },
                Object {
                  "label": "MT",
                  "value": "MT",
                },
                Object {
                  "label": "NE",
                  "value": "NE",
                },
                Object {
                  "label": "NV",
                  "value": "NV",
                },
                Object {
                  "label": "NH",
                  "value": "NH",
                },
                Object {
                  "label": "NJ",
                  "value": "NJ",
                },
                Object {
                  "label": "NM",
                  "value": "NM",
                },
                Object {
                  "label": "NY",
                  "value": "NY",
                },
                Object {
                  "label": "NC",
                  "value": "NC",
                },
                Object {
                  "label": "ND",
                  "value": "ND",
                },
                Object {
                  "label": "OH",
                  "value": "OH",
                },
                Object {
                  "label": "OK",
                  "value": "OK",
                },
                Object {
                  "label": "OR",
                  "value": "OR",
                },
                Object {
                  "label": "PA",
                  "value": "PA",
                },
                Object {
                  "label": "RI",
                  "value": "RI",
                },
                Object {
                  "label": "SC",
                  "value": "SC",
                },
                Object {
                  "label": "SD",
                  "value": "SD",
                },
                Object {
                  "label": "TN",
                  "value": "TN",
                },
                Object {
                  "label": "TX",
                  "value": "TX",
                },
                Object {
                  "label": "UT",
                  "value": "UT",
                },
                Object {
                  "label": "VT",
                  "value": "VT",
                },
                Object {
                  "label": "VA",
                  "value": "VA",
                },
                Object {
                  "label": "WA",
                  "value": "WA",
                },
                Object {
                  "label": "WV",
                  "value": "WV",
                },
                Object {
                  "label": "WI",
                  "value": "WI",
                },
                Object {
                  "label": "WY",
                  "value": "WY",
                },
              ],
            }
          }
        />
        <TextField
          id="formOne_AddressOne_zip"
          label="Zip"
          name="textField"
          onChange={[Function]}
          options={
            Object {
              "inputProps": Object {
                "className": "w-50",
              },
            }
          }
          placeholder=""
        />
      </ForwardRef>,
      "_rendering": false,
      "_updater": [Circular],
      "_workInProgressHook": null,
    },
  },
}
`;
